batch/batch/batch.py

```diff
@@ -998,10 +998,58 @@ def is_successful(self):
 
 
 async def _cancel_batch(batch_id, user):
-    batch = await Batch.from_db(batch_id, user)
-    if not batch:
-        abort(404)
-    await batch.cancel()
+    async with db.pool.acquire() as conn:
+        async with conn.cursor() as cursor:
+            await cursor.execute('''
+    update jobs
+inner join batch on jobs.batch_id == batch.batch_id
+       set jobs.state = 'Cancelled'
+     where jobs.batch_id = %s
+       and jobs.user = %s
+       and batch.deleted = FALSE
+       and batch.cancelled = FALSE
+       and jobs.state not in ('Cancelled', 'Error', 'Failed', 'Success', 'Pending')
+       and jobs.always_run == FALSE
+''', (batch_id, user))
+            await cursor.execute('''
+    update jobs
+inner join (    select jobs.id
+                  from jobs
+            inner join `jobs-parents` on `jobs-parents`.batch_id = jobs.batch_id
+                                     and `jobs-parents`.job_id = jobs.job_id
+                 where jobs.state = 'Pending'
+                   and jobs.batch_id = %s
+                   and jobs.user = %s
+                   and batch.deleted = FALSE
+                   and batch.cancelled = FALSE
+              group by jobs.job_id
+                having count(`jobs-parents`.state = 'Success') = count(*)
+                    or (jobs.always_run = TRUE and
+                        count(`jobs.parents`.state in ('Success', 'Failed', 'Cancelled')) = count(*))
+           ) as jobs2 on jobs2.job_id = jobs.job_id and jobs2.batch_id = jobs.batch_id
+       set state = 'Ready'
+''', (batch_id, user))
+            await cursor.execute('''
+update batch set cancelled = TRUE, closed = TRUE
+''')
+            await cursor.execute('''
+    select jobs.*
+      from jobs
+     where jobs.state = 'Ready' or jobs.state = 'Cancelled'
+       and jobs.batch_id = %s and user = %s
+       and batch.deleted = FALSE
+       and batch.cancelled = FALSE
+;
+''', (batch_id, user))
+            actionable_jobs = await cursor.fetchall()
+    for record in actionable_jobs:
+        job = Job.from_record(record)
+        if record['state'] == 'Cancelled':
+            await job._delete_k8s_resources()
```

Comment:
I think we should try and do bulk deletes. k8s has `delete_collection_namespaced_pod` and `delete_collection_namespaced_persistent_volume_claim`

Reply:
We should look into it. I think you can only specify a label selector. Maybe if we had always_run as a label then we could delete all the pods that are always_run=False and in the matching batch.